tests/models/test_dag.py

```diff
@@ -3524,3 +3519,259 @@ def test_create_dagrun_disallow_manual_to_use_automated_run_id(run_id_type: DagR
+    Here we verify, primarily with regard to setups and teardowns, the
+    behavior of dag.partial_subset but also the supporting methods defined
+    on AbstractOperator.
+    """
+
+    @staticmethod
+    def make_tasks(dag, input_str):
+        """
+        Helper for building setup and teardown tasks for testing.
+
+        Given an input such as 's1, w1, t1, tf1', returns setup task "s1", normal task "w1"
+        (the w means *work*), teardown task "t1", and teardown task "tf1" where the f means
+        on_failure_fail_dagrun has been set to true.
+        """
+
+        def teardown_task(task_id):
+            return BaseOperator.as_teardown(task_id=task_id)
+
+        def teardown_task_f(task_id):
+            return BaseOperator.as_teardown(task_id=task_id, on_failure_fail_dagrun=True)
+
+        def work_task(task_id):
+            return BaseOperator(task_id=task_id)
+
+        def setup_task(task_id):
+            return BaseOperator.as_setup(task_id=task_id)
+
+        def make_task(task_id):
+            """
+            Task factory helper.
+
+            Will give a setup, teardown, work, or teardown-with-dagrun-failure task depending on input.
+            """
+            if task_id.startswith("s"):
+                factory = setup_task
+            elif task_id.startswith("w"):
+                factory = work_task
+            elif task_id.startswith("tf"):
+                factory = teardown_task_f
+            elif task_id.startswith("t"):
+                factory = teardown_task
+            else:
+                raise ValueError("unexpected")
+            return dag.task_dict.get(task_id) or factory(task_id=task_id)
+
+        return (make_task(x) for x in input_str.split(", "))
+
+    @staticmethod
+    def cleared_downstream(task):
+        """Helper to return tasks that would be cleared if **downstream** selected."""
+        upstream = False
+        return set(
+            task.dag.partial_subset(
+                task_ids_or_regex=[task.task_id],
+                include_downstream=not upstream,
+                include_upstream=upstream,
+            ).tasks
+        )
+
+    @staticmethod
+    def cleared_upstream(task):
+        """Helper to return tasks that would be cleared if **upstream** selected."""
+        upstream = True
+        return set(
+            task.dag.partial_subset(
+                task_ids_or_regex=[task.task_id],
+                include_downstream=not upstream,
+                include_upstream=upstream,
+            ).tasks
+        )
+
+    def test_get_flat_relative_ids_with_setup(self):
+        with DAG(dag_id="test_dag", start_date=pendulum.now()) as dag:
+            s1, w1, w2, w3, t1 = self.make_tasks(dag, "s1, w1, w2, w3, t1")
+
+        s1 >> w1 >> w2 >> w3
+
+        # there is no teardown downstream of w2, so we assume w2 does not need s1
+        assert set(w2.get_upstreams_only_setups_and_teardowns()) == set()
```

Comment:
I feel like we should check that w1 does consider s1 a setup here before we add any other relationships.

Reply:
Yeah this was a point of ambiguity that I was meaning to bring up.  For now the behavior is you need to throw on an "empty" teardown in order to make the setup's scope well-defined.  @potiuk seemed to like that behavior.  the alternative is if a setup doesn't have a teardown then all its downstream tasks are assumed to be in its scope.   and that's definitely feasible and i don't really mind either way.  lmkyt

Reply:
Works for me, thanks for the explanation.

Reply:
i made the change already, so now the downstreams of a teardownless setup are all in the scope of that setup, so no empty teardown required. what do you think is best behavior?

Reply:
Sorry, I follow now. I don't have a strong preference either way to be honest. Both are clunky in their own way. But if we go back the original, we should check that w1 considers s1 a setup.

Reply:
original is, if we don't add empty teardown, then w1 would *not* consider s1 a setup unless s1 >> t1 and t1 downstream of w1.
so i think what we'd want to check in that scenario is that w1 does *not* consider s1 a setup

Reply:
Daniel and I had a chat and decided to go with the empty teardown approach for defining scope.